File: NegativeCutter-135.lrplugin/tests/test_direction_align_fourway.py
#!/usr/bin/env python3
"""
Four-way EXIF orientation mapping — isolated unit test (Python port)
Validates directionAlign rotation formulas for AB/BC/CD/DA orientations.

Since lua is not installed in this environment, this Python port verifies
the exact same math that the Lua four-way mapper would use.

Usage: python3 tests/test_direction_align_fourway.py
"""

import sys

pass_count = 0
fail_count = 0


def assert_eq(expected, actual, msg="", eps=1e-9):
    if isinstance(expected, tuple):
        if len(expected) != len(actual):
            raise AssertionError(
                f"{msg}: tuple length mismatch {len(expected)} vs {len(actual)}"
            )
        for i, (e, a) in enumerate(zip(expected, actual)):
            assert_eq(e, a, f"{msg}[{i}]", eps)
        return
    if abs(expected - actual) > eps:
        raise AssertionError(
            f"{msg}: expected {expected!r}, got {actual!r}"
        )


def run_test(name, fn):
    global pass_count, fail_count
    print(f"  {name} ... ", end="")
    try:
        fn()
        print("PASS")
        pass_count += 1
    except Exception as e:
        print(f"FAIL\n    {e}")
        fail_count += 1


# ------------------------------------------------------------------
# Stecman-derived formulas (relative coords 0-1, AB = normal)
# ------------------------------------------------------------------
def rotate_bc(t, b, l, r):
    """90 deg CW: (t,b,l,r) -> (1-r, 1-l, t, b)"""
    return 1.0 - r, 1.0 - l, t, b


def rotate_cd(t, b, l, r):
    """180 deg: (t,b,l,r) -> (1-b, 1-t, 1-r, 1-l)"""
    return 1.0 - b, 1.0 - t, 1.0 - r, 1.0 - l


def rotate_da(t, b, l, r):
    """270 deg CW / 90 deg CCW: (t,b,l,r) -> (l, r, 1-b, 1-t)"""
    return l, r, 1.0 - b, 1.0 - t


print("\n--- Test Suite: rotation formula group axioms ---")


def test_bc_roundtrip():
    t, b, l, r = 0.1, 0.4, 0.2, 0.8
    t1, b1, l1, r1 = rotate_bc(t, b, l, r)
    t2, b2, l2, r2 = rotate_bc(t1, b1, l1, r1)
    t3, b3, l3, r3 = rotate_bc(t2, b2, l2, r2)
    t4, b4, l4, r4 = rotate_bc(t3, b3, l3, r3)
    assert_eq(t, t4, "top")
    assert_eq(b, b4, "bottom")
    assert_eq(l, l4, "left")
    assert_eq(r, r4, "right")


def test_cd_roundtrip():
    t, b, l, r = 0.1, 0.4, 0.2, 0.8
    t1, b1, l1, r1 = rotate_cd(t, b, l, r)
    t2, b2, l2, r2 = rotate_cd(t1, b1, l1, r1)
    assert_eq(t, t2, "top")
    assert_eq(b, b2, "bottom")
    assert_eq(l, l2, "left")
    assert_eq(r, r2, "right")


def test_da_roundtrip():
    t, b, l, r = 0.1, 0.4, 0.2, 0.8
    t1, b1, l1, r1 = rotate_da(t, b, l, r)
    t2, b2, l2, r2 = rotate_da(t1, b1, l1, r1)
    t3, b3, l3, r3 = rotate_da(t2, b2, l2, r2)
    t4, b4, l4, r4 = rotate_da(t3, b3, l3, r3)
    assert_eq(t, t4, "top")
    assert_eq(b, b4, "bottom")
    assert_eq(l, l4, "left")
    assert_eq(r, r4, "right")


def test_bc_bc_equals_cd():
    t, b, l, r = 0.1, 0.4, 0.2, 0.8
    bc_bc = rotate_bc(*rotate_bc(t, b, l, r))
    cd = rotate_cd(t, b, l, r)
    assert_eq(cd, bc_bc, "combined")


def test_bc_cd_equals_da():
    t, b, l, r = 0.1, 0.4, 0.2, 0.8
    bc_cd = rotate_bc(*rotate_cd(t, b, l, r))
    da = rotate_da(t, b, l, r)
    assert_eq(da, bc_cd, "combined")


def test_cd_bc_equals_da():
    t, b, l, r = 0.1, 0.4, 0.2, 0.8
    cd_bc = rotate_cd(*rotate_bc(t, b, l, r))
    da = rotate_da(t, b, l, r)
    assert_eq(da, cd_bc, "combined")


print("\n--- Test Suite: geometric sanity checks ---")


def test_bc_centred_square():
    t, b, l, r = 0.1, 0.9, 0.2, 0.8
    nt, nb, nl, nr = rotate_bc(t, b, l, r)
    assert_eq(0.2, nt, "top")
    assert_eq(0.8, nb, "bottom")
    assert_eq(0.1, nl, "left")
    assert_eq(0.9, nr, "right")


def test_cd_centred_square():
    t, b, l, r = 0.1, 0.9, 0.2, 0.8
    nt, nb, nl, nr = rotate_cd(t, b, l, r)
    assert_eq(0.1, nt, "top")
    assert_eq(0.9, nb, "bottom")
    assert_eq(0.2, nl, "left")
    assert_eq(0.8, nr, "right")


def test_da_centred_square():
    t, b, l, r = 0.1, 0.9, 0.2, 0.8
    nt, nb, nl, nr = rotate_da(t, b, l, r)
    assert_eq(0.2, nt, "top")
    assert_eq(0.8, nb, "bottom")
    assert_eq(0.1, nl, "left")
    assert_eq(0.9, nr, "right")


def test_bc_preserves_aspect_ratio():
    t, b, l, r = 0.1, 0.5, 0.0, 1.0
    nt, nb, nl, nr = rotate_bc(t, b, l, r)
    orig_w = r - l
    orig_h = b - t
    new_w = nr - nl
    new_h = nb - nt
    assert_eq(orig_w, new_h, "width -> height")
    assert_eq(orig_h, new_w, "height -> width")


def test_da_preserves_aspect_ratio():
    t, b, l, r = 0.1, 0.5, 0.0, 1.0
    nt, nb, nl, nr = rotate_da(t, b, l, r)
    orig_w = r - l
    orig_h = b - t
    new_w = nr - nl
    new_h = nb - nt
    assert_eq(orig_w, new_h, "width -> height")
    assert_eq(orig_h, new_w, "height -> width")


print("\n--- Test Suite: edge cases ---")


def test_full_frame_bc():
    t, b, l, r = 0.0, 1.0, 0.0, 1.0
    nt, nb, nl, nr = rotate_bc(t, b, l, r)
    assert_eq(0.0, nt, "top")
    assert_eq(1.0, nb, "bottom")
    assert_eq(0.0, nl, "left")
    assert_eq(1.0, nr, "right")


def test_degenerate_width_bc():
    t, b, l, r = 0.1, 0.3, 0.3, 0.3
    nt, nb, nl, nr = rotate_bc(t, b, l, r)
    assert_eq(0.7, nt, "top")
    assert_eq(0.7, nb, "bottom")
    assert_eq(0.1, nl, "left")
    assert_eq(0.3, nr, "right")


# Run all tests
run_test("BC round-trip (4x = identity)", test_bc_roundtrip)
run_test("CD round-trip (2x = identity)", test_cd_roundtrip)
run_test("DA round-trip (4x = identity)", test_da_roundtrip)
run_test("BC + BC == CD", test_bc_bc_equals_cd)
run_test("BC + CD == DA", test_bc_cd_equals_da)
run_test("CD + BC == DA", test_cd_bc_equals_da)
run_test("BC on centred square frame", test_bc_centred_square)
run_test("CD on centred square frame", test_cd_centred_square)
run_test("DA on centred square frame", test_da_centred_square)
run_test("Full-strip frame BC preserves aspect ratio", test_bc_preserves_aspect_ratio)
run_test("Full-strip frame DA preserves aspect ratio", test_da_preserves_aspect_ratio)
run_test("Zero-margin frame BC", test_full_frame_bc)
run_test("Degenerate-width frame BC", test_degenerate_width_bc)

print("\n" + "-" * 62)
print(f"结果: {pass_count} 通过, {fail_count} 失败")
print("-" * 62)

if fail_count > 0:
    sys.exit(1)
else:
    print("\n全部测试通过!")

File: NegativeCutter-135.lrplugin/tests/test_direction_align_fourway.lua
#!/usr/bin/env lua
--[[
  Four-way EXIF orientation mapping — isolated unit test
  Validates directionAlign with AB / BC / CD / DA orientations.

  Usage: lua tests/test_direction_align_fourway.lua
]]--

local thisDir = debug.getinfo(1).source:match("@?(.*/)") or "./"
local pluginDir = thisDir .. "../"

package.path = package.path
    .. ";" .. thisDir .. "?.lua"
    .. ";" .. thisDir .. "mock_sdk/?.lua"
    .. ";" .. pluginDir .. "?.lua"

_G._PLUGIN = { path = pluginDir }

local mockSdk = require("mock_sdk.init")
mockSdk.setupImport()

-- Stub dofile for ThumbnailAgent / ApplierAgent
local _orig_dofile = dofile
_G.dofile = function(path)
    if type(path) == "string" and path:match("ThumbnailAgent%.lua$") then
        return { extract = function(_, _, cb) if cb then cb(true, "/tmp/t.jpg", nil) end end }
    end
    if type(path) == "string" and path:match("ApplierAgent%.lua$") then
        return {
            applyCrop = function() return true end,
            resetCrop = function() return true end,
        }
    end
    return _orig_dofile(path)
end

_G.LOC = function(key) return key end

local ProcessAgent = dofile(pluginDir .. "ProcessAgent.lua")
local MockPhoto = require("mock_sdk.LrPhoto")

local passCount, failCount = 0, 0

local function assertEq(expected, actual, msg)
    if expected ~= actual then
        error(string.format("%s: expected %s, got %s", msg or "assertEq",
            tostring(expected), tostring(actual)))
    end
end

local function runTest(name, fn)
    io.write("  " .. name .. " ... ")
    local ok, err = pcall(fn)
    if ok then print("PASS"); passCount = passCount + 1
    else print("FAIL\n    " .. tostring(err):gsub("\n", "\n    ")); failCount = failCount + 1 end
end

-- ------------------------------------------------------------------
-- Helper: create a result frame with explicit relative coords
-- ------------------------------------------------------------------
local function makeResult(opts)
    opts = opts or {}
    return {
        frameCount = 1,
        sourceWidth = opts.srcW or 2400,
        sourceHeight = opts.srcH or 3500,
        isHorizontal = (opts.srcW or 2400) >= (opts.srcH or 3500),
        cropAngle = opts.angle or 0,
        frames = {
            {
                index = 1,
                relativeTop = opts.rt or 0.1,
                relativeBottom = opts.rb or 0.4,
                relativeLeft = opts.rl or 0.0,
                relativeRight = opts.rr or 1.0,
            }
        }
    }
end

print("\n--- Test Suite: directionAlign baseline (existing behaviour) ---")

runTest("AB→AB (no rotation)", function()
    local photo = MockPhoto.createMockPhoto({
        dimensions = {width = 2400, height = 3500},
        rawMetadata = {orientation = "AB"}
    })
    local result = makeResult({srcW=2400, srcH=3500, rt=0.1, rb=0.4, rl=0.0, rr=1.0})
    local a = ProcessAgent.directionAlign(result, photo)
    assertEq(0.1, a.frames[1].relativeTop, "top")
    assertEq(0.4, a.frames[1].relativeBottom, "bottom")
    assertEq(0.0, a.frames[1].relativeLeft, "left")
    assertEq(1.0, a.frames[1].relativeRight, "right")
end)

runTest("90° mismatch rotates (existing behaviour)", function()
    -- Python vertical, LR horizontal
    local photo = MockPhoto.createMockPhoto({
        dimensions = {width = 3500, height = 2400},
        rawMetadata = {orientation = "AB"}
    })
    local result = makeResult({srcW=2400, srcH=3500, rt=0.1, rb=0.4, rl=0.0, rr=1.0})
    local a = ProcessAgent.directionAlign(result, photo)
    assertEq(0.0, a.frames[1].relativeTop, "rotated top")
    assertEq(1.0, a.frames[1].relativeBottom, "rotated bottom")
    assertEq(0.1, a.frames[1].relativeLeft, "rotated left")
    assertEq(0.4, a.frames[1].relativeRight, "rotated right")
end)

print("\n--- Test Suite: four-way mapping formulas (stecman-derived) ---")

-- These tests exercise the four rotation formulas that would be added to
-- directionAlign.  They are written as pure coordinate transforms so we
-- can verify the math without touching ProcessAgent yet.

--[[
  Stecman formulas (relative coords, 0-1, AB = normal / no rotation):

  BC (90° CW):
    right = bottom
    bottom = 1 - left
    left = top
    top = 1 - right

  CD (180°):
    bottom = 1 - top
    left = 1 - right
    top = 1 - bottom
    right = 1 - left

  DA (270° CW / 90° CCW):
    left = 1 - bottom
    top = left_original
    right = 1 - top_original
    bottom = right_original

  Wait — the DA formulas as written in stecman's code use the ORIGINAL
  crop values on the RHS.  To express them as transforms on the result:

  BC:  (t,b,l,r) → (1-r, 1-l, t, b)
  CD:  (t,b,l,r) → (1-b, 1-t, 1-r, 1-l)
  DA:  (t,b,l,r) → (l, r, 1-b, 1-t)

  Let's verify these compose correctly:
    AB → BC: (t,b,l,r) → (1-r, 1-l, t, b)
    BC → CD: (1-r, 1-l, t, b) → (1-(1-l), 1-(1-r), 1-b, 1-t)
             = (l, r, 1-b, 1-t)  ✓ matches DA from AB
    BC → BC: (1-r, 1-l, t, b) → (1-b, 1-t, 1-r, 1-l)
             = CD from AB  ✓

  So the clean transform matrix is:
    AB: identity
    BC: (t,b,l,r) ↦ (1-r, 1-l,  t,  b)
    CD: (t,b,l,r) ↦ (1-b, 1-t, 1-r, 1-l)
    DA: (t,b,l,r) ↦ ( l,   r, 1-b, 1-t)
--]]

local function rotateBC(t, b, l, r)
    return 1 - r, 1 - l, t, b
end

local function rotateCD(t, b, l, r)
    return 1 - b, 1 - t, 1 - r, 1 - l
end

local function rotateDA(t, b, l, r)
    return l, r, 1 - b, 1 - t
end

runTest("BC rotation formula round-trip", function()
    local t, b, l, r = 0.1, 0.4, 0.0, 1.0
    -- AB → BC
    local t1, b1, l1, r1 = rotateBC(t, b, l, r)
    -- BC → CD (= AB → BC → BC)
    local t2, b2, l2, r2 = rotateBC(t1, b1, l1, r1)
    -- CD → DA (= AB → BC → BC → BC)
    local t3, b3, l3, r3 = rotateBC(t2, b2, l2, r2)
    -- DA → AB (= 4×BC)
    local t4, b4, l4, r4 = rotateBC(t3, b3, l3, r3)

    assertEq(t, t4, "top after 4×BC")
    assertEq(b, b4, "bottom after 4×BC")
    assertEq(l, l4, "left after 4×BC")
    assertEq(r, r4, "right after 4×BC")
end)

runTest("CD rotation formula round-trip", function()
    local t, b, l, r = 0.1, 0.4, 0.2, 0.8
    local t1, b1, l1, r1 = rotateCD(t, b, l, r)
    local t2, b2, l2, r2 = rotateCD(t1, b1, l1, r1)
    assertEq(t, t2, "top after 2×CD")
    assertEq(b, b2, "bottom after 2×CD")
    assertEq(l, l2, "left after 2×CD")
    assertEq(r, r2, "right after 2×CD")
end)

runTest("DA rotation formula round-trip", function()
    local t, b, l, r = 0.1, 0.4, 0.2, 0.8
    local t1, b1, l1, r1 = rotateDA(t, b, l, r)
    local t2, b2, l2, r2 = rotateDA(t1, b1, l1, r1)
    local t3, b3, l3, r3 = rotateDA(t2, b2, l2, r2)
    local t4, b4, l4, r4 = rotateDA(t3, b3, l3, r3)
    assertEq(t, t4, "top after 4×DA")
    assertEq(b, b4, "bottom after 4×DA")
    assertEq(l, l4, "left after 4×DA")
    assertEq(r, r4, "right after 4×DA")
end)

runTest("BC+BC equals CD", function()
    local t, b, l, r = 0.1, 0.4, 0.2, 0.8
    local bc_bc_t, bc_bc_b, bc_bc_l, bc_bc_r = rotateBC(rotateBC(t, b, l, r))
    local cd_t, cd_b, cd_l, cd_r = rotateCD(t, b, l, r)
    assertEq(cd_t, bc_bc_t, "top")
    assertEq(cd_b, bc_bc_b, "bottom")
    assertEq(cd_l, bc_bc_l, "left")
    assertEq(cd_r, bc_bc_r, "right")
end)

runTest("BC+CD equals DA", function()
    local t, b, l, r = 0.1, 0.4, 0.2, 0.8
    local bc_cd_t, bc_cd_b, bc_cd_l, bc_cd_r = rotateBC(rotateCD(t, b, l, r))
    local da_t, da_b, da_l, da_r = rotateDA(t, b, l, r)
    assertEq(da_t, bc_cd_t, "top")
    assertEq(da_b, bc_cd_b, "bottom")
    assertEq(da_l, bc_cd_l, "left")
    assertEq(da_r, bc_cd_r, "right")
end)

runTest("BC on a centred square frame", function()
    -- A centred frame: top=0.1, bottom=0.9, left=0.2, right=0.8
    -- After BC: top=0.2, bottom=0.8, left=0.1, right=0.9
    -- Visual: frame rotates 90° CW, so narrow side becomes height
    local t, b, l, r = 0.1, 0.9, 0.2, 0.8
    local nt, nb, nl, nr = rotateBC(t, b, l, r)
    assertEq(0.2, nt, "top")
    assertEq(0.8, nb, "bottom")
    assertEq(0.1, nl, "left")
    assertEq(0.9, nr, "right")
end)

runTest("CD on a centred square frame", function()
    -- 180° flip should mirror around centre
    local t, b, l, r = 0.1, 0.9, 0.2, 0.8
    local nt, nb, nl, nr = rotateCD(t, b, l, r)
    assertEq(0.1, nt, "top (mirrored)")
    assertEq(0.9, nb, "bottom (mirrored)")
    assertEq(0.2, nl, "left (mirrored)")
    assertEq(0.8, nr, "right (mirrored)")
end)

runTest("DA on a centred square frame", function()
    -- 90° CCW: top=0.2, bottom=0.8, left=0.1, right=0.9
    local t, b, l, r = 0.1, 0.9, 0.2, 0.8
    local nt, nb, nl, nr = rotateDA(t, b, l, r)
    assertEq(0.2, nt, "top")
    assertEq(0.8, nb, "bottom")
    assertEq(0.1, nl, "left")
    assertEq(0.9, nr, "right")
end)

print("\n--- Test Suite: cropAngle rotation sign ---")

runTest("BC negates cropAngle once", function()
    -- angle changes sign on every 90° rotation
    local a = 1.5
    assertEq(-1.5, -a, "negated angle")
end)

runTest("CD preserves cropAngle (2×negation)", function()
    local a = 1.5
    assertEq(1.5, -(-a), "double negation")
end)

print("\n" .. string.rep("-", 62))
print(string.format("结果: %d 通过, %d 失败", passCount, failCount))
print(string.rep("-", 62))

if failCount > 0 then os.exit(1)
else print("\n全部测试通过!") end
